File: 212.Word-Search-II.py
# https://leetcode.com/problems/word-search-ii/
#
# algorithms
# Hard (27.72%)
# Total Accepted:    100,953
# Total Submissions: 364,212
# beats 61.83% of python submissions


class TrieNode(object):
    def __init__(self):
        self.children = [None] * 26
        self.is_leaf = False


class Solution(object):
    def findWords(self, board, words):
        """
        :type board: List[List[str]]
        :type words: List[str]
        :rtype: List[str]
        """
        trie_root = TrieNode()
        row, col = len(board), len(board[0])

        for w in words:
            tmp = trie_root
            for ch in w:
                trie_idx = ord(ch) - 97
                if tmp.children[trie_idx] is None:
                    tmp.children[trie_idx] = TrieNode()
                tmp = tmp.children[trie_idx]
            tmp.is_leaf = True

        def recursive(i, j, root, path):
            ch = board[i][j]

            if ch == '#':
                return

            trie_idx = ord(ch) - 97
            if root.children[trie_idx] is None:
                return

            path += ch
            board[i][j] = '#'
            if root.children[trie_idx].is_leaf:
                res[0] += path,
                root.children[trie_idx].is_leaf = False

            if i > 0:
                recursive(i - 1, j, root.children[trie_idx], path)

            if i < row - 1:
                recursive(i + 1, j, root.children[trie_idx], path)

            if j > 0:
                recursive(i, j - 1, root.children[trie_idx], path)

            if j < col - 1:
                recursive(i, j + 1, root.children[trie_idx], path)

            path = path[:-1]
            board[i][j] = ch

        res = [[]]

        for i in xrange(row):
            for j in xrange(col):
                    recursive(i, j, trie_root, '')

        return res[0]

File: 212.Word-Search-II.go
// https://leetcode.com/problems/word-search-ii/
//
// algorithms
// Hard (28.27%)
// Total Accepted:    108,169
// Total Submissions: 382,655
// beats 81.01% of golang submissions

package leetcode

type TrieNode struct {
	word     string
	children []*TrieNode
}

var res []string
var isVisited [][]bool

func findWords(board [][]byte, words []string) []string {
	trieTree := buildTrieTree(words)

	res = []string{}
	row, col := len(board), len(board[0])
	isVisited = make([][]bool, row)
	for i := 0; i < row; i++ {
		isVisited[i] = make([]bool, col)
	}

	for i := 0; i < row; i++ {
		for j := 0; j < col; j++ {
			isVisited[i][j] = true
			recursive(board, i, j, trieTree)
			isVisited[i][j] = false
		}
	}

	return res
}

func recursive(board [][]byte, i, j int, node *TrieNode) {
	row, col := len(board), len(board[0])
	idx := (int)(board[i][j] - 'a')

	if node.children[idx] == nil {
		return
	}

	node = node.children[idx]
	if node.word != "" {
		res = append(res, node.word)
		node.word = ""
	}

	if i > 0 && !isVisited[i-1][j] {
		isVisited[i-1][j] = true
		recursive(board, i-1, j, node)
		isVisited[i-1][j] = false
	}

	if i < row-1 && !isVisited[i+1][j] {
		isVisited[i+1][j] = true
		recursive(board, i+1, j, node)
		isVisited[i+1][j] = false
	}

	if j > 0 && !isVisited[i][j-1] {
		isVisited[i][j-1] = true
		recursive(board, i, j-1, node)
		isVisited[i][j-1] = false
	}

	if j < col-1 && !isVisited[i][j+1] {
		isVisited[i][j+1] = true
		recursive(board, i, j+1, node)
		isVisited[i][j+1] = false
	}

}

func buildTrieTree(words []string) *TrieNode {
	root := &TrieNode{
		children: make([]*TrieNode, 26),
	}

	for _, w := range words {
		tmp := root
		for _, ch := range w {
			idx := (int)(ch - 'a')
			if tmp.children[idx] == nil {
				tmp.children[idx] = &TrieNode{
					children: make([]*TrieNode, 26),
				}
			}
			tmp = tmp.children[idx]
		}
		tmp.word = w
	}

	return root
}
